### ISSUE TITLE
Add "runfile_path" to File

### ISSUE THREAD

**Issue by pauldraper** [bazelbuild/bazel#14307]
### Description of the problem / feature request:

Add `runfile_path` property to `File` object.

### What underlying problem are you trying to solve with this feature?

Runfiles are identified by path including their workspace.

`short_path` is the location of the file relative to the current working directory, when the executable is run under Bazel.

However, that doesn't work if the executable is run outside of Bazel. For that, the file must be found in the runfiles manifest, or the runfiles symlink tree. However, neither of these use the same root as the cwd. Namely, the cwd directory is $runfiles/$workspace_name.

So path manipulation is required. And it requires `ctx.workspace_name()` which isn't available in `Args.add_all()` `map_each`.

Ideally, I could simply use `runfile_path` and get the path to use in runfile lookup/rlocation.

**Comment by fmeum** [comment]
This would be very useful, but I could see `rlocation_path` being a name that's less overloaded.

Regarding `Args.add_all()`: Bazel 5 has support for nested functions and lambdas on Starlark, so you can close over the value of `ctx.workspace_name()` in you function passed to `map_each`.

**Comment by pauldraper** [comment]
> Bazel 5 has support for nested functions and lambdas in Starlark

Yes, though it has to be explicitly enabled because the cost is (apparently) significant.

**Comment by thii** [comment]
> the cost is (apparently) significant.

@pauldraper Do you know where can I find more details about this?

> it has to be explicitly enabled

Also, I'm not sure if I'm missing something but we're actually using them in the Swift rules without having to explicitly enable anything.

**Comment by pauldraper** [comment]
> allow_closure
> default = False
> If true, allows the use of closures in function parameters like map_each. Usually this isn't necessary and it risks retaining large analysis-phase data structures into the execution phase.

https://docs.bazel.build/versions/main/skylark/lib/Args.html#add_all

**Comment by pauldraper** [comment]
Regardless of the support for lambdas, there would be significant utility in providing `runfile_path` out-of-the-box.

Unless rule authors are 100% certain the users will never depend on one executable from another, `short_path` by itself is useless and even harmful.

**Comment by fmeum** [comment]
This will be very easy to implement after WORKSPACE has been removed.

**Comment by sputt** [comment]
Thanks for reopening. This issue is a source of friction when folks are learning to author rules. "So if it's an executable I should use file.short_path?" "Well yes, but technically no."

I like File.rlocation_path

**Comment by fmeum** [comment]
@meteorcloudy @Wyverald Now that this is trivial to implement as a struct field on `FileApi`, what do you think about adding it?

**Comment by meteorcloudy** [comment]
SGTM, but @Wyverald should confirm ;)

**Comment by Silic0nS0ldier** [comment]
@Wyverald Yay or nay on this?

Currently making do with this bandaid;

```starlark
# runfiles.bzl
def runfiles_path(label, include_name = True):
    if type(label) == "string":
        label = Label(label)
    path_parts = [label.repo_name if label.repo_name != "" else "_main"]
    if label.package != "":
        path_parts.append(label.package)
    if include_name:
        path_parts.append(label.name)
    return "/".join(path_parts)
# BUILD.bazel
load("@aspect_bazel_lib//lib:expand_template.bzl", "expand_template")
load(":runfiles.bzl", "runfiles_path")
expand_template(
    name = "cst_config",
    substitutions = { "{{foo_runfiles_path}}": runfiles_path("@foo//:bin/foo") },
    template = "container-test.yaml",
)
```

This _works_ but is quite brittle (label is unchecked), so keen for proper support in the Bazel API to better justify refactoring for the ~166 usages.

**Comment by fmeum** [comment]
@Silic0nS0ldier That's exactly the use case for `$(rlocationpath ...)`.

**Comment by Silic0nS0ldier** [comment]
`$(rlocationpath ...)` has some problems;
- Can only reference targets (to address their default outputs) and `attr.output(_list)` attributes.
- Because only targets can be used, it is necessary use a custom rule (or a `filegroup` if an appropriate output group is defined) to give a label to the outputs needed.

Additionally for `$(rlocationpaths ...)`
- It is unsafe to use with file paths containing spaces, as the result is a space delimited string.

Not a good foundation to build on. Even for the simple `expand_template` example, we have many inputs which would require a helper to pick the desired file.

**Comment by pauldraper** [comment]
> This will be very easy to implement after WORKSPACE has been removed.

🎉

**Comment by pauldraper** [comment]
This available in a library as `to_rlocation_path(ctx, file)` in bazel_lib:

https://registry.bazel.build/docs/bazel_lib#function-to_rlocation_path

**Comment by phst** [comment]
Also see https://github.com/bazelbuild/bazel/issues/17259
